### PR TITLE
Refactor part 3

### PR DESCRIPTION
**Pull request by lehmann-4178656ch**
This pull tries to merge all logic needed to determine step execution order for all relevant actions (rm, build, pull, run) into one single function. This reduces the possibility for introducing errors when changing commands or adding new ones.

### CODE REVIEW

File: pipeline.go [hunk 1]
```diff
@@ -345,59 +345,96 @@ func (p *PipelineDefinition) Pipelines() (*Pipelines, error) {
...
+}
 
-	select {
-	case <-abort:
+func runCommandParallel(config runConfig, runner Runner, step Step, durations *sync.Map, wg *sync.WaitGroup, preconditions []chan struct{}, done chan struct{}, abort chan error) {
+	defer wg.Done()
+	defer close(done)
+	for i, c := range preconditions {
+		if Verbose {
+			pipelineLogger.Printf("%s waiting for %d preconditions", step.ColoredContainerName(), len(preconditions)-i)
```
Comment (niklas88) [inline pipeline.go]: Shouldn't this be the singluar in the Printf as it prints the number of the current precondition?

File: pipeline.go [hunk 2]
```diff
@@ -345,59 +345,96 @@ func (p *PipelineDefinition) Pipelines() (*Pipelines, error) {
 	return p.pipelines, nil
 }
 
-func runParallelBuildImage(runner Runner, step Step, pull bool, durations *sync.Map, wg *sync.WaitGroup, start chan struct{}, abort chan struct{}) {
-	defer wg.Done()
-	<-start
+type runConfig struct {
+	usePreconditions bool
```
Comment (niklas88) [inline pipeline.go]: Can't this information be encoded as just a `nil` `pre` func?

File: pipeline.go [hunk 3]
```diff
@@ -406,133 +443,106 @@ func (p Pipeline) BuildImages(force bool) error {
...
-			}
-			if err := runner.ContainerRemover(step)(); err != nil {
-				pipelineLogger.Printf("Error removing %s: %s", step.ColoredName(), err)
+	return p.runCommand(runConfig{
+		usePreconditions: false,
+		selection: func(step Step) bool {
+			return !preRun || step.Meta.KeepAlive != KeepAliveReplace
+		},
+		pre: func(runner Runner, step Step) error {
+			return nil
```
Comment (niklas88) [inline pipeline.go]: As above I feel like it would be more intuitive for `pre` itself to be `nil`

**Comment by niklas88** [comment]
Overall I again really like this change. Combining this into one concept absolutely makes sense.

File: pipeline.go (hunk 2 above)
Comment (lehmann-4178656ch) [reply pipeline.go]: Yes it can and will be, I just didn't think about that.

File: pipeline.go (hunk 3 above)
Comment (lehmann-4178656ch) [reply pipeline.go]: Will be changed and simplified.

File: pipeline.go (hunk 1 above)
Comment (lehmann-4178656ch) [reply pipeline.go]:
```
x waiting for 1 precondition
x waiting for 1 precondition(s)
x waiting for 1 preconditions

x waiting for 4 precondition
x waiting for 4 precondition(s)
x waiting for 4 preconditions
```
I'd suggest adding the `()`s to the message or did you mean a different place?

**Comment by lehmann-4178656ch** [comment]
Thanks for the suggestions, I've added them in 3d8a7d7 and had to add some more lines in e3c7710 to restore the old output, without e3c7710 all steps print `Execute:` and how long it took "run steps" regardless of the read work done.

**Comment by niklas88** [comment]
Just wanted to say that the code looks good to me but I would still like to give it another test compile and try it out if this isn't too time critical.

**Comment by lehmann-4178656ch** [comment]
Thank you for your time!

**Pull request merged**